addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/ThirdGeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,767 @@
+    // base URL of the web api
+    private static final String WEB_API = "/api/v1";
+
+    /*
+     * operations used for authentication
+     */
+    private static final String AUTH_START = "/auth/start";
+    private static final String AUTH_FINISH = "/auth/finish";
+    private static final String AUTH_CREATE_SESSION = "/auth/create_session";
+
+    /*
+     * operations used for gathering process data from the device
+     */
+    private static final String PROCESSDATA = "/processdata";
+
+    /*
+     * List of available channels for the concrete device
+     */
+
+    /*
+     * After the authentication the result (the session id) is stored here and used to "sign" future requests
+     */
+    @Nullable
+    protected String sessionId = null;
+    /*
+     * The configuration file containing the host, the password and the refresh interval
+     */
+    @Nullable
+    private ThirdGenerationConfiguration config;
+
+    @Nullable
+    protected ScheduledFuture<?> refreshScheduler;
+
+    /**
+     * Constructor of this class
+     *
+     * @param thing the thing
+     */
+    public ThirdGenerationHandler(Thing thing, ThirdGenerationInverterTypes inverterType) {
+        super(thing);
+        this.inverterType = inverterType;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        /*
+         * All channels are readonly and updated by the scheduler
+         */
+    }
+
+    @Override
+    public void dispose() {
+        if (refreshScheduler != null) {
+            refreshScheduler.cancel(true);
+        }
+        super.dispose();
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public void initialize() {
+        config = getConfigAs(ThirdGenerationConfiguration.class);
+        /*
+         * temporary value while initializing
+         */
+        updateStatus(ThingStatus.UNKNOWN);
+
+        /*
+         * Start the initialization
+         */
+        authenticate();
+
+        /*
+         * Start the update scheduler as configured
+         */
+        refreshScheduler = scheduler.scheduleWithFixedDelay(() -> {
+            updateChannelValues();
+        }, 10, config.refreshInternalInSeconds, TimeUnit.SECONDS);
+    }
+
+    /**
+     * The API supports the resolution of multiple values at a time
+     *
+     * Therefore this methods builds one request to gather all information for the current inverter.
+     * The list contains all channels as defined in {@link ThirdGenerationMappingInverterToChannel} for the
+     * current inverter
+     *
+     */
+    void updateChannelValues() {
+        Map<String, List<ThirdGenerationChannelMappingToWebApi>> channelList = ThirdGenerationMappingInverterToChannel
+                .getModuleToChannelsMappingForInverter(inverterType);
+        JsonArray updateMessageJsonArray = new JsonArray();
+        /*
+         * Build the message to send to the inverter
+         */
+        for (String moduleId : channelList.keySet()) {
```

Comment:
Prefer to iterate over `entrySet` as `keySet` makes a copy of the keys every time called.

Reply:
will be in the next commit